integration/simulation/network/obscuro_node_utils.go

```diff
@@ -61,11 +63,34 @@ func startInMemoryObscuroNodes(params *params.SimParams, stats *stats.Stats, gen
 		obscuroClients[i] = host.NewInMemObscuroClient(node)
 	}
 	time.Sleep(100 * time.Millisecond)
-	return obscuroClients
+
+	walletClients := setupWalletClientsWithoutViewingKeys(params, obscuroClients)
+
+	return obscuroClients, walletClients
+}
+
+// setupWalletClientsWithoutViewingKeys will configure the existing obscuro clients to be used as wallet clients,
+// (we typically keep a client per wallet so their viewing keys are available and registered but they can share clients
+// 	if no viewing keys are in use)
+func setupWalletClientsWithoutViewingKeys(params *params.SimParams, obscuroClients []rpcclientlib.Client) map[string]rpcclientlib.Client {
+	walletClients := make(map[string]rpcclientlib.Client)
+	var i int
+	// loop through all the L2 wallets we're using and round-robin allocate them the rpc clients we have for each host
+	for _, w := range params.Wallets.SimObsWallets {
+		walletClients[w.Address().String()] = obscuroClients[i%len(obscuroClients)]
+		i++
+	}
+	for _, t := range params.Wallets.Tokens {
+		w := t.L2Owner
+		walletClients[w.Address().String()] = obscuroClients[i%len(obscuroClients)]
+		i++
+	}
+	return walletClients
 }
 
-func startStandaloneObscuroNodes(params *params.SimParams, stats *stats.Stats, gethClients []ethadapter.EthClient, enclaveAddresses []string) []rpcclientlib.Client {
+func startStandaloneObscuroNodes(params *params.SimParams, stats *stats.Stats, gethClients []ethadapter.EthClient, enclaveAddresses []string) ([]rpcclientlib.Client, map[string]rpcclientlib.Client) {
```

Comment:
What if the starting an obscuro node and creating RPC Clients / Wallets was decoupled ?

Reply:
I agree in principle but I've just had a go and it's fiddly to make it come out cleanly (like you might need to pass the addresses back as well as the nodes). I think maybe this is a prime spot for a refactor PR (have left a todo).